packages/repo/src/structure.ts

```diff
@@ -0,0 +1,289 @@
+    return this.updateRepo({
+      stagedWrites: [...this.stagedWrites, ...writeArr],
+    })
+  }
+
+  async createCommit(
+    authStore: auth.AuthStore,
+    performUpdate?: (prev: CID, curr: CID) => Promise<CID | null>,
+  ): Promise<RepoStructure> {
+    let data = this.data
+    for (const write of this.stagedWrites) {
+      if (write.action === 'create') {
+        const dataKey = write.collection + '/' + write.rkey
+        data = await data.add(dataKey, write.cid)
+      } else if (write.action === 'update') {
+        const dataKey = write.collection + '/' + write.rkey
+        data = await data.update(dataKey, write.cid)
+      } else if (write.action === 'delete') {
+        const dataKey = write.collection + '/' + write.rkey
+        data = await data.delete(dataKey)
+      }
+    }
+    const token = (await authStore.canSignForDid(this.did()))
+      ? null
+      : await util.ucanForOperation(this.data, data, this.did(), authStore)
+    const tokenCid = token ? await this.blockstore.put(token) : null
+
+    const dataCid = await data.save()
+    const root: RepoRoot = {
+      meta: this.root.meta,
+      prev: this.cid,
+      auth_token: tokenCid,
+      data: dataCid,
+    }
+    const rootCid = await this.blockstore.put(root)
+    const commit: Commit = {
+      root: rootCid,
+      sig: await authStore.sign(rootCid.bytes),
+    }
+    const commitCid = await this.blockstore.put(commit)
+
+    if (performUpdate) {
+      const rebaseOn = await performUpdate(this.cid, commitCid)
+      if (rebaseOn) {
+        const rebaseRepo = await RepoStructure.load(this.blockstore, rebaseOn)
+        return rebaseRepo.createCommit(authStore, performUpdate)
+      }
+    }
+
+    return this.updateRepo({
+      cid: commitCid,
+      root,
+      commit,
+      data,
+      stagedWrites: [],
+    })
+  }
+
+  async revert(count: number): Promise<RepoStructure> {
+    let revertTo = this.cid
+    for (let i = 0; i < count; i++) {
+      const commit = await this.blockstore.get(revertTo, def.commit)
+      const root = await this.blockstore.get(commit.root, def.repoRoot)
+      if (root.prev === null) {
+        throw new Error(`Could not revert ${count} commits`)
+      }
+      revertTo = root.prev
+    }
+    return RepoStructure.load(this.blockstore, revertTo)
+  }
+
+  // CAR FILES
+  // -----------
+
+  async getCarNoHistory(): Promise<Uint8Array> {
+    return this.openCar((car: BlockWriter) => {
+      return this.writeCheckoutToCarStream(car)
+    })
+  }
+
+  async getDiffCar(to: CID | null): Promise<Uint8Array> {
+    return this.openCar((car: BlockWriter) => {
+      return this.writeCommitsToCarStream(car, to, this.cid)
+    })
+  }
+
+  async getFullHistory(): Promise<Uint8Array> {
+    return this.getDiffCar(null)
+  }
+
+  private async openCar(
+    fn: (car: BlockWriter) => Promise<void>,
+  ): Promise<Uint8Array> {
+    const { writer, out } = CarWriter.create([this.cid])
+    await fn(writer)
+    writer.close()
```

Comment:
I wonder if this could cause some kind of leak without a little extra love:
```suggestion
    try {
      await fn(writer)
    } finally {
      writer.close()
    }
```

Reply:
ah excellent. thanks for the catch 🙏